File: 747.largest-number-at-least-twice-of-others.cpp
/*
 * @lc app=leetcode id=747 lang=cpp
 *
 * [747] Largest Number At Least Twice of Others
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  int dominantIndex(vector<int>& nums) {
    int size = nums.size();
    if (size < 2) {
      return 0;
    }
    int largest = INT_MIN;
    int largest_idx = -1;
    int second = INT_MIN;
    int second_idx = -1;
    for (int i = 0; i < size; i++) {
      if (nums[i] >= largest) {
        second = largest;
        second_idx = largest_idx;
        largest = nums[i];
        largest_idx = i;
      } else if (nums[i] > second) {
        second = nums[i];
        second_idx = i;
      }
    }
    return (largest >= 2 * second) ? largest_idx : -1;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> nums;
  nums = {3};
  ASSERT_EQ(s.dominantIndex(nums), 0);
  nums = {3, 6, 1, 0};
  ASSERT_EQ(s.dominantIndex(nums), 1);
  nums = {1, 2, 3, 4};
  ASSERT_EQ(s.dominantIndex(nums), -1);
  return 0;
}

File: 747.largest-number-at-least-twice-of-others.rs
/*
 * @lc app=leetcode id=747 lang=rust
 *
 * [747] Largest Number At Least Twice of Others
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn dominant_index(nums: Vec<i32>) -> i32 {
        let size = nums.len();
        if size < 2 {
            return 0;
        }
        let mut largest = i32::MIN;
        let mut largest_idx: i32 = -1;
        let mut second = i32::MIN;
        for i in 0..size {
            if nums[i] >= largest {
                second = largest;
                largest = nums[i];
                largest_idx = i as i32;
            } else if nums[i] > second {
                second = nums[i];
            }
        }
        if largest as i64 >= 2 * second as i64 {
            largest_idx
        } else {
            -1
        }
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::dominant_index(vec![3, 6, 1, 0]), 1);
}

#[test]
fn test2() {
    assert_eq!(Solution::dominant_index(vec![1, 2, 3, 4]), -1);
}
